File: tests/adapters/cpp/goldens/do/function_bodies.cpp
/**
 * C++ module for testing function body optimization.
 */

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

struct User {
    int id;
    std::string name;
    std::string email;
};

struct ProcessingResult {
    std::vector<User> valid;
    std::vector<User> invalid;
};

class Calculator {
private:
    std::vector<std::string> history;
    std::string name;

public:
    Calculator() : Calculator("default") {}

    Calculator(const std::string& name) : name(name) {
        history.reserve(10);
    }

    int add(int a, int b) {
        int result = a + b;
        history.push_back("add(" + std::to_string(a) + ", " + std::to_string(b) + ") = " + std::to_string(result));
        std::cout << "Addition result: " << result << std::endl;
        return result;
    }

    int multiply(int a, int b) {
        int result = a * b;
        history.push_back("multiply(" + std::to_string(a) + ", " + std::to_string(b) + ") = " + std::to_string(result));
        return result;
    }

    std::vector<std::string> getHistory() const {
        return history;
    }

private:
    bool validateInput(int value) {
        std::string str = std::to_string(value);

        for (char c : str) {
            if (c != '-' && (c < '0' || c > '9')) {
                std::cerr << "Input must be a number" << std::endl;
                return false;
            }
        }

        if (value == INT_MAX || value == INT_MIN) {
            std::cerr << "Input must be finite" << std::endl;
            return false;
        }

        return true;
    }
};

ProcessingResult processUserData(const std::vector<User>& users) {
    ProcessingResult result;

    for (const auto& user : users) {
        if (user.id > 0 &&
            !user.name.empty() &&
            user.email.find('@') != std::string::npos) {
            result.valid.push_back(user);
        } else {
            result.invalid.push_back(user);
        }
    }

    return result;
}

// Template function
template<typename T, typename Processor>
std::vector<T> processArray(const std::vector<T>& items, Processor processor) {
    std::vector<T> result;

    for (const auto& item : items) {
        try {
            result.push_back(processor(item));
        } catch (const std::exception& error) {
            std::cerr << "Processing failed for item" << std::endl;
        }
    }

    return result;
}

// Lambda usage function
std::vector<int> filterPositive(const std::vector<int>& numbers) {
    std::vector<int> result;
    std::copy_if(numbers.begin(), numbers.end(), std::back_inserter(result),
                 [](int n) { return n > 0; });
    return result;
}

int main() {
    Calculator calc("test");
    std::cout << calc.add(2, 3) << std::endl;
    std::cout << calc.multiply(4, 5) << std::endl;

    std::vector<User> users = {
        {1, "Alice", "alice@example.com"},
        {2, "Bob", "bob@example.com"}
    };

    ProcessingResult processed = processUserData(users);
    std::cout << "Valid users: " << processed.valid.size() << std::endl;

    return 0;
}

File: tests/adapters/rust/goldens/do/function_bodies.rs
// Rust module for testing function body optimization.

use std::collections::HashMap;

#[derive(Debug, Clone)]
struct User {
    id: i32,
    name: String,
    email: String,
}

#[derive(Debug)]
struct ProcessingResult {
    valid: Vec<User>,
    invalid: Vec<User>,
}

struct Calculator {
    name: String,
    history: Vec<String>,
}

impl Calculator {
    fn new() -> Self {
        Self::with_name("default")
    }

    fn with_name(name: &str) -> Self {
        Self {
            name: name.to_string(),
            history: Vec::with_capacity(10),
        }
    }

    fn add(&mut self, a: i32, b: i32) -> i32 {
        let result = a + b;
        let entry = format!("add({}, {}) = {}", a, b, result);
        self.history.push(entry);
        println!("Addition result: {}", result);
        result
    }

    fn multiply(&mut self, a: i32, b: i32) -> i32 {
        let result = a * b;
        let entry = format!("multiply({}, {}) = {}", a, b, result);
        self.history.push(entry);
        result
    }

    fn get_history(&self) -> Vec<String> {
        self.history.clone()
    }

    fn validate_input(&self, value: i32) -> bool {
        let value_str = value.to_string();

        for ch in value_str.chars() {
            if ch != '-' && !ch.is_ascii_digit() {
                eprintln!("Input must be a number");
                return false;
            }
        }

        if value == i32::MAX || value == i32::MIN {
            eprintln!("Input must be finite");
            return false;
        }

        true
    }
}

fn process_user_data(users: Vec<User>) -> ProcessingResult {
    let mut result = ProcessingResult {
        valid: Vec::new(),
        invalid: Vec::new(),
    };

    for user in users {
        if user.id > 0 && !user.name.is_empty() && user.email.contains('@') {
            result.valid.push(user);
        } else {
            result.invalid.push(user);
        }
    }

    result
}

fn process_array<T, F>(items: Vec<T>, processor: F) -> Vec<T>
where
    F: Fn(T) -> T,
{
    let mut result = Vec::new();

    for item in items {
        let processed = processor(item);
        result.push(processed);
    }

    result
}

fn filter_positive(numbers: Vec<i32>) -> Vec<i32> {
    numbers.into_iter().filter(|&n| n > 0).collect()
}

fn main() {
    let mut calc = Calculator::with_name("test");
    println!("{}", calc.add(2, 3));
    println!("{}", calc.multiply(4, 5));

    let users = vec![
        User {
            id: 1,
            name: "Alice".to_string(),
            email: "alice@example.com".to_string(),
        },
        User {
            id: 2,
            name: "Bob".to_string(),
            email: "bob@example.com".to_string(),
        },
    ];

    let processed = process_user_data(users);
    println!("Valid users: {}", processed.valid.len());
}
